streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopologyBuilder.java

```diff
@@ -1455,27 +1447,46 @@ public void addSuccessor(final TopologyDescription.Node successor) {
     }
 
     public final static class Source extends AbstractNode implements TopologyDescription.Source {
-        private final String topics;
+        private final Set<String> topics;
+        private final Pattern topicPattern;
 
         public Source(final String name,
-                      final String topics) {
+                      final Set<String> topics,
+                      final Pattern pattern) {
             super(name);
             this.topics = topics;
+            this.topicPattern = pattern;
         }
 
+        @Deprecated
         @Override
         public String topics() {
+            return topics.toString();
+        }
+
+        @Override
+        public Set<String> topicSet() {
             return topics;
         }
 
+        @Override
+        public Pattern topicPattern() {
+            return topicPattern;
+        }
+
         @Override
         public void addPredecessor(final TopologyDescription.Node predecessor) {
             throw new UnsupportedOperationException("Sources don't have predecessors.");
         }
 
         @Override
         public String toString() {
-            return "Source: " + name + " (topics: " + topics + ")\n      --> " + nodeNames(successors);
+            if (topics == null) {
```

Comment:
nit: you could just specify one `String` variable at the top of the method and set it accordingly if `topics` is `null` or not then have a single `return` statement.  Just a personal preference though.